File: library/py/FenwickTree.py
class FenwickTree:

    def __init__(self, n):
        self.size = n
        self.tree = [0] * self.size

    def add(self, index, value):
        x = index
        while x < self.size:
            self.tree[x] += value
            x |= x + 1

    def sum(self, index):
        # 半開区間[0,index)
        ret = 0
        x = index - 1
        while x >= 0:
            ret += self.tree[x]
            x = (x & (x + 1)) - 1
        return ret

File: library/cpp/FenwickTree.cc
template <typename T>
class FenwickTree {
private:
    vector<T> tree;
public:
    FenwickTree(int size) : tree(size, 0) {
    }
    void add(int index, T value){
    	while(index < tree.size()) {
    		tree[index] += value;
    		index |= index + 1;
    	}
    }
    T sum(int index){
    	T res = 0;
    	--index;
    	while(index >= 0) {
    		res += tree[index];
    		index = (index & (index + 1)) - 1;
    	}
    	return res;
    }
};
